### ISSUE TITLE
do not block pes that are sending messages due to other messaging activity

### ISSUE THREAD

**Issue by jcphill** [UIUC-PPL/charm#642]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/642*

---
Attached is a trace of PME in NAMD for netlrts-linux-x86_64-smp-tcp running two processes on a single node, showing the PME sendUntrans entry points in yellow with inter-node messages traced. There is clearly some kind of lock in the messaging layer, as when an inter-node send is attempted the call blocks waiting for the completion of both other outgoing message sends from the same node but also incoming message receives. A less severe effect is seen on the PME sendTrans entry points, also in yellow at the left edge of the image. A udp (non-tcp) build has the same issue.

![comm_server_user_events_fixed](https://user-images.githubusercontent.com/1772435/63628306-5c5dd200-c5d1-11e9-8266-1277efc27a95.png)
![comm_server_user_events](https://user-images.githubusercontent.com/1772435/63628307-5c5dd200-c5d1-11e9-8c2a-d165f5856e49.png)
![namd_user_events_lrts_send](https://user-images.githubusercontent.com/1772435/63628308-5cf66880-c5d1-11e9-8dd8-30492db8057d.png)
![netlrts-linux-x86_64-smp-tcp](https://user-images.githubusercontent.com/1772435/63628309-5cf66880-c5d1-11e9-9f51-1f3e7b659562.png)
![withoutprogresscalls](https://user-images.githubusercontent.com/1772435/63628310-5cf66880-c5d1-11e9-9059-76b09193074c.png)

**Comment by jcphill** [comment]
*Original date: 2014-12-31 18:28:47*

---
This is the situation after removing CmiMachineProgressImpl() calls from the pes.  Sending entry methods are clearly delayed by incoming messages.

**Comment by jcphill** [comment]
*Original date: 2014-12-31 19:34:07*

---
Looking at the code, this is a case of the send functions calling CmiCommLock() on the pe threads in order to interface directly with the network.  This is silly since the communication thread will be idle during this period, waiting for the communication lock.  If only one thread is going to make progress, it might as well be the communication thread.  Pes that want to send a message should drop the message into the communication thread's outgoing queue, which should be as fast as delivering a message to another pe on the same node.  Then the communication thread can manage all sends and receives collectively, ideally overlapping send and receive, and also paying attention to the assigned message priorities.

**Comment by jcphill** [comment]
*Original date: 2015-02-09 19:59:12*

---
Ping.

**Comment by bilgeacun** [comment]
*Original date: 2015-02-19 17:07:17*

---
Marking it as bug.

**Comment by jcphill** [comment]
*Original date: 2015-03-04 18:59:25*

---
Ping.

**Comment by bilgeacun** [comment]
*Original date: 2015-03-05 19:46:10*

---
This commit http://charm.cs.uiuc.edu/gerrit/#/c/577/  should solve the problem. 

Worker threads were unnecessarily calling the communication server from CmiMachineProgressImpl function. I have added a check to make sure it's only called from the comm thread.

I'm attaching two projections images which is showing the problem. I have added trace user events in the communication server when the thread is holding the communication lock. In comm_server_user_events.png, light blue bars above the entry methods should not exist for worker threads. 
in the fixed version comm_server_user_events_fixed.png, the light blue bars do not exist for the worker threads.

**Comment by jcphill** [comment]
*Original date: 2015-03-06 17:03:15*

---
That may also be an issue, but the problem I was reporting is that LrtsSendFunc takes the comm lock to call DeliverOutgoingMessage().
DeliverOutgoingMessage() also takes the comm lock internally (if LrtsSendFunc() didn't already) and calls DeliverViaNetwork().  (It appears to do nothing to on-node messages, as those are delivered by CmiPushPE() in CmiInterFreeSendFn().)
DeliverViaNetwork() assumes it holds the comm lock and calls EnqueueOutgoingDgram.

The problem is that this is the same comm lock that CommunicationServerNet holds while looping over CheckSocketsReady() and its other communication functions, causing any thread that is trying to push a message into the outgoing queue to hang.  Pushing a message into the comm thread outgoing queue should be similar to CmiSendNodeSelf() or CmiPushPE(), which use LRTSQueuePush() (which uses either lockless or a dedicated lock rather than one big node lock).

Basically, the communication thread should only hold a lock to its own outgoing message queue long enough to read or modify it, not during the time it takes to actually transmit the data.

**Comment by jcphill** [comment]
*Original date: 2015-03-06 17:15:47*

---
I see that CommunicationServer() for non-LRTS net and ibverbs have the same lock-holding behavior.  This is likely a problem for all communication layers.

**Comment by bilgeacun** [comment]
*Original date: 2015-03-06 19:06:50*

---
Yes, I was suspecting that this is the problem as well. To confirm that this is the problem, I have added trace user events around LrtsSend function and the problem can be clearly seen from the namd_user_events_lrts_send.png image that I've just added. 
The yellow entry methods are PME send_trans functions and the green bars over them shows the time they spend in LrtsSend. When yellow entries take too much time(more than the other yellows), most of the time is spent in LrtsSend and this is because comm thread is holding the comm lock(this can be seen from the light blue bars over the comm thread).

I'm implementing a fix for this. My earlier commit on gerrit is not solving this problem but this problem revealed another problem related to CmiMachineProgressImpl at least.
